Closes #632

rohangpta opened the pull request:
- Add logic to interact with the CyberSource API to validate transaction data and also confirm the payment.
- Validate the transient token's integrity using the public key API.
- Add appropriate error handling for API invocation failures causing transaction failure.
- Store the transaction data in a new model `TicketTransactionRecord` for bookkeeping purposes. Each ticket is also associated with an instance of this class.
- On transaction success, assign the ticket to the user, remove holds and from cart, and send out confirmation email.

rm03 reviewed (approved):
lgtm, just a few minor comments

rohangpta commented on backend/clubs/models.py:
```diff
@@ -1814,6 +1814,18 @@ def update_holds(self):
+    """
+
+    total_amount = models.FloatField()
+    buyer_phone = PhoneNumberField(null=True, blank=True)
```
Are they required by all payment methods?

rm03 replied:
I think so? At least in the iframe, it didn't let me continue unless I filled out those fields

rohangpta replied:
I added `null=True` cos wasn't sure if  other payment methods like Apple Pay, bank transfer etc required email input. Can you check with the iframe if they do and let me know?

rm03 replied:
I don't think there's any easy way to check other payment methods with the test API, can leave this as is for now.

rohangpta commented on backend/clubs/views.py:
```diff
@@ -4778,47 +4817,95 @@ def checkout_success_callback(self, request, *args, **kwargs):
+        except ApiException as e:
+            # Cleanup state since the purchase failed
+            for ticket in cart.tickets.select_for_update().all():
+                ticket.holder = None
```
My thinking:

if `initiate_checkout` fails -- card details entered incorrectly or something of that sort... user error (HTTP 400). Don't clear holds and give another chance to the user.

if `complete_checkout` fails -- something very bad happened; we got a transient token but we couldn't auth using it. (HTTP 500). We have a dangling transient token now: maybe the service will fail to checkout again if we input the card details, or it's down entirely. Either way, it's out of our control and we should clear holds...

I'll add status codes to responses make this clear. I should have done that to begin with :)

rm03 replied:
Sorry I might be misunderstanding, but isn't `initiate_checkout` completely independent of the user's card details, since we just take their cart total and generate a capture context? 

In `complete_checkout`, if the user's entered their card details incorrectly, then the authorization will fail and that transient token is now useless. Even if the authorization is successful, the capture could still fail if the bank flags the transaction as fraudulent.

However I'm thinking it might be better to maintain the hold and send them back to the cart page so they can restart checkout with new payment details.

rm03 replied:
Resolved offline

rohangpta replied:
Resolved offline:

Transient tokens represent a complete payment authorization after validating payment details and funds in someone's account for a charge. So, if a payment fails after getting a transient token, then something truly went wrong...which means nuking holds is expected.

aviupadhyayula reviewed (commented):
🚀

rohangpta commented on backend/clubs/views.py:
```diff
@@ -4778,47 +4817,95 @@ def checkout_success_callback(self, request, *args, **kwargs):
+                                    type: boolean
         ---
         """
+        tt = request.data.get("transient_token")
```
Yep adding that in as we speak! Thanks

rohangpta commented on backend/clubs/views.py:
```diff
@@ -4778,47 +4817,95 @@ def checkout_success_callback(self, request, *args, **kwargs):
-        """
+        except ApiException as e:
+            # Cleanup state since the purchase failed
+            for ticket in cart.tickets.select_for_update().all():
+                ticket.holder = None
+                ticket.owner = None
+                ticket.save()
```
Yes, will refactor here and other places.

rohangpta commented on backend/clubs/views.py:
```diff
@@ -4778,47 +4817,95 @@ def checkout_success_callback(self, request, *args, **kwargs):
+            ).get_transaction_for_transient_token(tt)
 
-        return Response({"detail": "callback successful"})
+            if not transaction_data or (status % 100) in [4, 5]:
```
Sure

rohangpta commented on backend/clubs/views.py:
```diff
@@ -4778,47 +4817,95 @@ def checkout_success_callback(self, request, *args, **kwargs):
+
+        # At this point, we have validated that the payment was authorized
+        # Give the tickets to the user
+        for ticket in cart.tickets.select_for_update(holder=self.request.user).all():
+            ticket.owner = request.user
+            ticket.carts.remove(cart)
+            ticket.holder = None
+            ticket.transaction_record = transaction_record
+            ticket.save()
+            ticket.send_confirmation_email()
```
Sure, I think there's some redundancy here:
```py
Ticket.objects.filter(id__in=[ticket.id for ticket in tickets]) == tickets
```

rohangpta commented:
@rm03 @aviupadhyayula can I get a final stamp on this? I updated a bunch of the logic in the most recent 2 commits, including transient token verification logic and a few other minor changes.

I tested this locally using the procedure I dropped in #penn-clubs-active. The entire flow from `initiate_checkout` to `complete_checkout` works.

aviupadhyayula commented on backend/clubs/views.py:
```diff
@@ -2336,29 +2370,38 @@ def add_to_cart(self, request, *args, **kwargs):
+                    {"detail": f"Not enough tickets of type {type} left!"},
+                    status=status.HTTP_403_FORBIDDEN,
+                )
+            else:
+                cart.tickets.add(*tickets[:count])
```
nit: can we use the guard clause pattern here

aviupadhyayula commented on backend/clubs/views.py:
```diff
@@ -2487,7 +2537,11 @@ def tickets(self, request, *args, **kwargs):
             available.append(
                 {
                     "type": type,
-                    "count": (tickets.filter(type=type, owner__isnull=True).count()),
+                    "count": (
+                        tickets.filter(
+                            type=type, owner__isnull=True, holder__isnull=True
+                        ).count()
+                    ),
                 }
             )
```
Does something like this work instead? 
```
totals = tickets.values('type').annotate(count=Count('type')).order_by('type')
available = tickets.filter(owner__isnull=True, holder__isnull=True).values('type').annotate(count=Count('type')).order_by('type')
return Response({"totals": list(totals), "available": list(available)})
```

rohangpta replied:
Yep, seems good to me. I'll add it in, thanks - that's another TODO gone

aviupadhyayula reviewed (commented):
Left some comments, but they're not blockers. LGTM 🚀
